File: java/structure/linear/queue/DequeOfLinkedList.java
package structure.linear.queue;

public class DequeOfLinkedList {
    private ListNode _head = new ListNode(0);
    private ListNode _tail = new ListNode(0);

    public DequeOfLinkedList() {
        _head.next = _tail;
        _tail.prev = _head;
    }

    public boolean isEmpty() {
        return _head.next == _tail;
    }

    public void enqueueHead(int value) {
        ListNode node = new ListNode(value);
        node.next = _head.next;
        node.prev = _head;
        _head.next.prev = node;
        _head.next = node;
    }

    public void enqueueTail(int value) {
        ListNode node = new ListNode(value);
        node.prev = _tail.prev;
        node.next = _tail;
        _tail.prev.next = node;
        _tail.prev = node;
    }

    public int dequeueHead() {
        if (_head.next != _tail) {
            ListNode node = _head.next;
            node.next.prev = _head;
            _head.next = node.next;
            return node.value;
        }
        return Integer.MIN_VALUE;
    }

    public int dequeueTail() {
        if (_head.next != _tail) {
            ListNode node = _tail.prev;
            node.prev.next = _tail;
            _tail.prev = node.prev;
            return node.value;
        }
        return Integer.MIN_VALUE;
    }

    public void clear() {
        _head.next = _tail;
        _tail.prev = _head;
    }

    private class ListNode {
        public int value = 0;
        public ListNode prev = null;
        public ListNode next = null;

        public ListNode(int value) {
            this.value = value;
        }
    }
}

File: py/structure/linear/queue/DequeOfLinkedList.py
#!/usr/bin/env python
# -*- coding:gbk -*-

import sys

class DequeOfLinkedList:
    def __init__(self):
        self.__head, self.__tail = self.__ListNode(0), self.__ListNode(0)
        self.__head.next, self.__tail.prev = self.__tail, self.__head

    def isEmpty(self) -> bool:
        return self.__head.next == self.__tail

    def enqueueHead(self, value: int) -> None:
        node = self.__ListNode(value)
        node.next, node.prev, self.__head.next.prev, self.__head.next = self.__head.next, self.__head, node, node

    def enqueueTail(self, value: int) -> None:
        node = self.__ListNode(value)
        node.prev, node.next, self.__tail.prev.next, self.__tail.prev = self.__tail.prev, self.__tail, node, node

    def dequeueHead(self) -> int:
        if self.__head.next != self.__tail:
            node = self.__head.next
            node.next.prev, self.__head.next = self.__head, node.next
            return node.value
        return -sys.maxsize

    def dequeueTail(self) -> int:
        if self.__head.next != self.__tail:
            node = self.__tail.prev
            node.prev.next, self.__tail.prev = self.__tail, node.prev
            return node.value
        return -sys.maxsize

    def clear(self) -> None:
        self.__head.next, self.__tail.prev = self.__tail, self.__head

    class __ListNode:
        def __init__(self, value: int):
            self.value, self.prev, self.next = value, None, None
